crates/red_knot_server/src/server/api.rs

```diff
@@ -85,9 +84,10 @@ fn background_request_task<'a, R: traits::BackgroundDocumentRequestHandler>(
         let Ok(path) = url_to_system_path(&url) else {
             return Box::new(|_, _| {});
         };
-        let db = session
-            .workspace_db_for_path(path.as_std_path())
-            .map(RootDatabase::snapshot);
+        let db = match session.workspace_db_for_path(path.as_std_path()) {
+            Some(db) => db.snapshot(),
+            None => session.default_workspace_db().snapshot(),
+        };
```

Comment:
I want this entire logic in the `Session` but the borrow checker always complains when I do something like:
```rs
        self.workspaces
            .range_mut(..=path.as_ref().to_path_buf())
            .next_back()
            .map(|(_, db)| db)
			.unwrap_or_else(|| self.default_workspace_db_mut())
```
There are two mutable references:
```
error[E0500]: closure requires unique access to `*self` but it is already borrowed
   --> crates/red_knot_server/src/session.rs:104:29
    |
97  |           &mut self,
    |           - let's call the lifetime of this reference `'1`
...
100 |           self.workspaces
    |           ---------------
    |           |
    |  _________borrow occurs here
    | |
101 | |             .range_mut(..=path.as_ref().to_path_buf())
102 | |             .next_back()
103 | |             .map(|(_, db)| db)
104 | |             .unwrap_or_else(|| self.default_workspace_db_mut())
    | |_____________________________^^_----___________________________- returning this value requires that `self.workspaces` is borrowed for `'1`
    |                               |  |
    |                               |  second borrow occurs due to use of `*self` in closure
    |                               closure construction occurs here

For more information about this error, try `rustc --explain E0500`.
```

Reply:
But yeah, not sure why the `mut` case doesn't work. Sounds like a borrow checker limitation? Maybe @BurntSushi has an idea

I tried to with as many hints as possible but it just won't believe that we only return one mutable borrow 

```rust
        let workspace = self
            .workspaces
            .range_mut(..=path.as_ref().to_path_buf())
            .next_back();

        if let Some((_, db)) = workspace {
            return db;
        }
        drop(workspace);

        return self.workspaces.values_mut().next().unwrap();
```

Reply:
> Not sure if I'm doing something wrong. But I think you face the problem with `workspace_db_for_path_mut`?

Yes, it's for the mutable version.

Reply:
Thanks. Yes, that might be it. 

An alternative solution could be to accept a closure that applies the mutation to the `db`. Not as nice but  hopefully that works?

Reply:
> An alternative solution could be to accept a closure that applies the mutation to the `db`. Not as nice but hopefully that works?

Huh, this works but it gets a little bit complicated because we need to move the `path` to the `ChangedEvent`. I'll keep the existing logic for now with a TODO to link to this discussion.